File: Packages/OpenVibbleKit/Sources/BuddyPersona/PersonaState.swift
import Foundation

public enum PersonaState: UInt8, Sendable, CaseIterable {
    case sleep
    case idle
    case busy
    case attention
    case celebrate
    case dizzy
    case heart

    public var slug: String {
        switch self {
        case .sleep: return "sleep"
        case .idle: return "idle"
        case .busy: return "busy"
        case .attention: return "attention"
        case .celebrate: return "celebrate"
        case .dizzy: return "dizzy"
        case .heart: return "heart"
        }
    }
}

public struct PersonaDeriveInput: Sendable, Equatable {
    public let connected: Bool
    public let sessionsRunning: Int
    public let sessionsWaiting: Int
    public let recentlyCompleted: Bool

    public init(
        connected: Bool,
        sessionsRunning: Int,
        sessionsWaiting: Int,
        recentlyCompleted: Bool
    ) {
        self.connected = connected
        self.sessionsRunning = sessionsRunning
        self.sessionsWaiting = sessionsWaiting
        self.recentlyCompleted = recentlyCompleted
    }
}

public enum PersonaOverlay: Sendable, Equatable {
    case none
    case sleep(since: Date)
    case dizzy(until: Date)
    case heart(until: Date)
    case celebrate(until: Date)
}

public func derivePersonaState(_ input: PersonaDeriveInput) -> PersonaState {
    if input.recentlyCompleted { return .celebrate }
    if !input.connected { return .idle }
    if input.sessionsWaiting > 0 { return .attention }
    if input.sessionsRunning >= 3 { return .busy }
    if input.sessionsRunning >= 1 { return .busy }
    return .idle
}

public func resolvePersonaState(
    base: PersonaState,
    overlay: PersonaOverlay,
    now: Date
) -> PersonaState {
    switch overlay {
    case .dizzy(let until) where now < until:
        return .dizzy
    case .heart(let until) where now < until:
        return .heart
    case .sleep(let since) where now.timeIntervalSince(since) >= 3:
        return .sleep
    case .celebrate(let until) where now < until:
        return .celebrate
    default:
        return base
    }
}

File: android/buddy-persona/src/main/kotlin/com/openvibble/persona/PersonaState.kt
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

package com.openvibble.persona

enum class PersonaState(val slug: String) {
    SLEEP("sleep"),
    IDLE("idle"),
    BUSY("busy"),
    ATTENTION("attention"),
    CELEBRATE("celebrate"),
    DIZZY("dizzy"),
    HEART("heart");

    companion object {
        fun fromSlug(slug: String): PersonaState? = values().firstOrNull { it.slug == slug }
    }
}

data class PersonaDeriveInput(
    val connected: Boolean,
    val sessionsRunning: Int,
    val sessionsWaiting: Int,
    val recentlyCompleted: Boolean,
)

/** Time-bounded state overrides applied on top of the derived base state. */
sealed class PersonaOverlay {
    data object None : PersonaOverlay()
    data class Sleep(val sinceEpochMs: Long) : PersonaOverlay()
    data class Dizzy(val untilEpochMs: Long) : PersonaOverlay()
    data class Heart(val untilEpochMs: Long) : PersonaOverlay()
    data class Celebrate(val untilEpochMs: Long) : PersonaOverlay()
}

fun derivePersonaState(input: PersonaDeriveInput): PersonaState {
    if (input.recentlyCompleted) return PersonaState.CELEBRATE
    if (!input.connected) return PersonaState.IDLE
    if (input.sessionsWaiting > 0) return PersonaState.ATTENTION
    if (input.sessionsRunning >= 1) return PersonaState.BUSY
    return PersonaState.IDLE
}

fun resolvePersonaState(
    base: PersonaState,
    overlay: PersonaOverlay,
    nowEpochMs: Long,
): PersonaState = when (overlay) {
    is PersonaOverlay.Dizzy -> if (nowEpochMs < overlay.untilEpochMs) PersonaState.DIZZY else base
    is PersonaOverlay.Heart -> if (nowEpochMs < overlay.untilEpochMs) PersonaState.HEART else base
    is PersonaOverlay.Sleep ->
        if (nowEpochMs - overlay.sinceEpochMs >= SLEEP_GRACE_MS) PersonaState.SLEEP else base
    is PersonaOverlay.Celebrate -> if (nowEpochMs < overlay.untilEpochMs) PersonaState.CELEBRATE else base
    PersonaOverlay.None -> base
}

private const val SLEEP_GRACE_MS = 3_000L
